### PR TITLE
Functional Testing

### PR DESCRIPTION
**Pull request by jimrojerTNO**
This PR is meant to clean up our ci/cd test pipeline and to add tests where we think they are needed. Please also look at the excel sheet for tracing where what is tested. 

You will mainly find tests to make it more complete, furthermore you will find replacements of tests where hardcoded pycml/modellica was used before. Specifically for the buffer and for the test_heat.py.

Please note that the pressuredrop tests are not good yet, this will be done in a separate PR by Kobus later.

### CODE REVIEW

File: tests/test_heat.py [hunk 1]
```diff
@@ -169,16 +173,15 @@ def _get_min_max_pressure(case):
         min_, max_ = _get_min_max_pressure(case_min_max_pressure)
         self.assertGreater(min_, self.min_pressure * 0.99)
         self.assertLess(max_, self.max_pressure * 1.01)
-        self.assertGreater(case_min_max_pressure.objective_value, base_objective_value * 1.5)
```
Comment (FJanssen-TNO) [inline tests/test_heat.py:172]: why was this removed and not replaced by self.assertGreater(case_min_max_pressure.objective_value, base_objective_value)
or add check that the demand is thus not matched, basically what this objective_value should be.

File: tests/test_multiple_in_and_out_port_components.py [hunk 2]
```diff
@@ -88,9 +97,12 @@ def test_heat_pump(self):
         sec_heat = results["GenericConversion_3d3f.Secondary_heat"]
         power_elec = results["GenericConversion_3d3f.Power_elec"]
 
-        # TODO: check if the primary source utilisisation is maximised and secondary minimised
-        # I think it is badly scaled or some scaling issues in constraints. (possible in combination
-        # with disconnected)
+        # Check that only the minimum velocity is flowing through the secondary source.
+        np.testing.assert_allclose(
+            results["ResidualHeatSource_aec9.Q"] / (np.pi * (0.3127 / 2.0) ** 2),
```
Comment (FJanssen-TNO) [inline tests/test_multiple_in_and_out_port_components.py:102]: assign (np.pi * (0.3127 / 2.0) ** 2) to a variable describing what it is, namely cross sectional area

File: tests/test_multiple_in_and_out_port_components.py [hunk 3]
```diff
@@ -88,9 +97,12 @@ def test_heat_pump(self):
...
+            1.0e-3,
```
Comment (FJanssen-TNO) [inline tests/test_multiple_in_and_out_port_components.py:103]: is it possible to extract this from the parameters?

File: tests/test_producer_profiles.py [hunk 4]
```diff
@@ -17,12 +17,9 @@ class TestProducerMaxProfile(TestCase):
     (reducing the profile value at a few time steps).
 
     Checks:
-    - Demand matching
+    - Standard checks demand matching, energy conservation and heat to discharge
```
Comment (FJanssen-TNO) [inline tests/test_producer_profiles.py:20]: check at line 53 can be removed as it is part of the demand_matching_test

File: tests/test_setpoint_constraints.py [hunk 5]
```diff
@@ -65,11 +65,6 @@ def test_run_small_ates_timed_setpoints_2_changes(self):
         Checks:
         - That the setpoint does change twice over window length if 2 is specified
 
-        Missing:
-        - This test runs extremely slow in current master, should be fixed!
```
Comment (FJanssen-TNO) [inline tests/test_setpoint_constraints.py:69]: was this fixed?

File: tests/test_setpoint_constraints.py [hunk 6]
```diff
@@ -65,11 +65,6 @@ def test_run_small_ates_timed_setpoints_2_changes(self):
...
-        _ this test does not check for funtioning with varying timestep sizes, this or another test
-        should be adapted to do so or a seperate test for that should be created.
```
Comment (FJanssen-TNO) [inline tests/test_setpoint_constraints.py:71]: I don't see an adaptation for this or a new test

File: tests/test_warmingup_unit_cases.py [hunk 7]
```diff
@@ -72,6 +124,42 @@ def test_3a(self):
...
+            np.sign(results["Pipe_e53a.Q"][inds]),
+        )
+
+        for buffer in heat_problem.heat_network_components.get("buffer", []):
+            np.testing.assert_allclose(
+                results[f"{buffer}.Heat_buffer"], results[f"{buffer}.Heat_flow"]
+            )
+            np.testing.assert_allclose(
+                results[f"{buffer}.HeatIn.Heat"] - results[f"{buffer}.HeatOut.Heat"],
+                results[f"{buffer}.Heat_buffer"],
```
Comment (FJanssen-TNO) [inline tests/test_warmingup_unit_cases.py:148]: shouldn't we add this test to the utils_tests? we do have it in that list for the heat_demand check

File: tests/test_ates.py [hunk 8]
```diff
@@ -21,8 +21,6 @@ def test_ates(self):
         - cyclic storage behaviour
         - standard energy conservation, etc.
 
-        Missing checks:
-        - should check ates history for artificial energy at t=0
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_ates.py:25]: I added a line in the missing for the endscenariosizing workflow to do this. I think we should not do this in the history, but we should do something with the cyclic behaviour. That's why I for now deleted it over here

File: tests/test_buffer.py [hunk 9]
```diff
@@ -1,64 +0,0 @@
...
-
-class TestBufferHistory(TestCase):
-    def test_buffer_history(self):
-        """
-        Check the buffer asset and its interaction with the network.
-
-        Checks:
-        - The logic of how the history is set for the buffer. This is to avoid artificial energy
-        at t=0.
-        - Checks that when this is overwritten this artificial energy is indeed exploited.
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_buffer.py:17]: this is now in the warmingup_unit_cases.py in the test_3a, line 165 where we check that the first heat_buffer == 0. This is also what is actually done in the history method.

File: tests/test_electric_source_sink.py [hunk 10]
```diff
@@ -65,20 +66,44 @@ def test_source_sink(self):
...
+            np.testing.assert_allclose(
+                results[f"{demand}.ElectricityIn.V"] * results[f"{demand}.ElectricityIn.I"],
+                results[f"{demand}.ElectricityIn.Power"],
+                atol=1.0e-3,
+            )
+
     def test_source_sink_max_curr(self):
         """
-        Check bounds on the current.
+        Check bounds on the current, this is achieved by upping the demand forcing the current to
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_electric_source_sink.py:95]: done

File: tests/test_electric_source_sink.py [hunk 11]
```diff
@@ -87,11 +112,16 @@ def test_source_sink_max_curr(self):
...
 
         base_folder = Path(example.__file__).resolve().parent.parent
-        max_ = 32660  # This max is based on max current and voltage requirement at consumer
-        v_min = 230  # set as minimum voltage for cables
 
         solution = run_optimization_problem(ElectricityProblemMaxCurr, base_folder=base_folder)
         results = solution.extract_results()
+        parameters = solution.parameters(0)
+
+        max_ = (
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_electric_source_sink.py:120]: done

File: tests/test_electric_source_sink.py [hunk 12]
```diff
@@ -135,3 +165,15 @@ def test_source_sink_max_curr(self):
         np.testing.assert_allclose(current_cable, current_producer)
         biggerthen = all(142.0 * np.ones(len(current_demand)) >= current_demand - tolerance)
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_electric_source_sink.py:166]: done

File: tests/test_electrolyzer.py [hunk 13]
```diff
@@ -11,7 +11,16 @@
...
+        This test is to check the functioning the example with an offshore wind farm in combination
+        with an electrolyzer and hydrogen storage.
+
+        Checks:
+        - The objective value with the revenue included
+        - Check the bounds on the electrolyzer
+        - Check the setpoint for the windfarm
+        - Check the max production profile of the windfarm
+        - Check the electrolyzer inequality constraints formulation
+
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_electrolyzer.py:23]: agree, but these fall a bit out of scope for NWN. I want to improve this part once we are starting the work with Javier more

File: tests/test_gas_multi_demand_source_node.py
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_gas_multi_demand_source_node.py:1]: agree, but this falls out of scope for NWN for now

File: tests/test_heat.py (hunk 1 above)
Comment (jimrojerTNO) [reply tests/test_heat.py:172]: I thought this was a bit of a unnecessary check, but I made something to replace it. Please note that both problems do not match demand.

File: tests/test_multiple_in_and_out_port_components.py (hunk 2 above)
Comment (jimrojerTNO) [reply tests/test_multiple_in_and_out_port_components.py:102]: ok

File: tests/test_setpoint_constraints.py (hunk 5 above)
Comment (jimrojerTNO) [reply tests/test_setpoint_constraints.py:69]: Yes as in it now runs fast, but we did not completely debug what was the cause

File: tests/test_setpoint_constraints.py (hunk 6 above)
Comment (jimrojerTNO) [reply tests/test_setpoint_constraints.py:71]: This was actually already in, the timeline had varying timestep size

**Review by FJanssen-TNO — APPROVED** [review]

**Pull request merged**